File: epi_judge_java_solutions/epi/RookAttack.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class RookAttack {

  public static void rookAttack(List<List<Integer>> A) {

    int m = A.size(), n = A.get(0).size();
    boolean hasFirstRowZero = A.get(0).contains(0);
    boolean hasFirstColumnZero = A.stream().anyMatch(row -> row.get(0) == 0);

    for (int i = 1; i < m; ++i) {
      for (int j = 1; j < n; ++j) {
        if (A.get(i).get(j) == 0) {
          A.get(i).set(0, 0);
          A.get(0).set(j, 0);
        }
      }
    }

    for (int i = 1; i < m; ++i) {
      if (A.get(i).get(0) == 0) {
        Collections.fill(A.get(i), 0);
      }
    }

    for (int j = 1; j < n; ++j) {
      if (A.get(0).get(j) == 0) {
        final int idx = j;
        A.stream().forEach(row -> row.set(idx, 0));
      }
    }

    if (hasFirstRowZero) {
      Collections.fill(A.get(0), 0);
    }
    if (hasFirstColumnZero) {
      A.stream().forEach(row -> row.set(0, 0));
    }
  }

  @EpiTest(testDataFile = "rook_attack.tsv")
  public static List<List<Integer>> rookAttackWrapper(List<List<Integer>> A) {
    List<List<Integer>> aCopy = new ArrayList<>(A);
    rookAttack(aCopy);
    return aCopy;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "RookAttack.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python_solutions/rook_attack.py
import copy
from typing import List

from test_framework import generic_test


def rook_attack(A: List[List[int]]) -> None:

    m, n = len(A), len(A[0])
    has_first_row_zero = 0 in A[0]
    has_first_column_zero = any(not A[i][0] for i in range(m))

    for i in range(1, m):
        for j in range(1, n):
            if not A[i][j]:
                A[i][0] = A[0][j] = 0

    for i in range(1, m):
        if not A[i][0]:
            for j in range(1, n):
                A[i][j] = 0

    for j in range(1, n):
        if not A[0][j]:
            for i in range(1, m):
                A[i][j] = 0

    if has_first_row_zero:
        for j in range(n):
            A[0][j] = 0

    if has_first_column_zero:
        for i in range(m):
            A[i][0] = 0


def rook_attack_wrapper(A):
    a_copy = copy.deepcopy(A)
    rook_attack(a_copy)
    return a_copy


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('rook_attack.py', 'rook_attack.tsv',
                                       rook_attack_wrapper))
